scripts/db_migrations/v4.16/fixSequencesActivities.py

```diff
@@ -0,0 +1,153 @@
+      return fileContent
+    else:
+      return getFileGridFS(db, collection_name, id)
+
+  return getFileGridFS(db, collection_name,entry["link"])
+
+
+def removeFileFs(fsfolder, link):
+  filename = os.path.join(fsfolder, link)
+
+  if not os.path.isfile(filename):
+    return False
+
+  os.remove(filename)
+  return True
+
+def removeFileGridFs(db, collection_name, link):
+  fs = gridfs.GridFS(db, collection_name)
+  gridfsfile =  fs.find_one({ "filename": link })
+
+  if gridfsfile:
+    fs.delete(gridfsfile._id)
+
+def removeFile(db, fsfolder, collection_name, id):
+
+  entry = db[collection_name + ".ref"].find_one({"_id": id})
+
+  if entry:
+    if entry["type"] == "fs":
+      if removeFileFs(fsfolder, entry["link"]):
+        print "File succesfully removed from filesystem" + entry["link"]
+      else:
+        print "File unsuccesfully removed from filesystem" + entry["link"]
+
+    removeFileGridFs(db, collection_name, entry["link"])
+  else:
+    print "Fileref not found " + id
+
+  db[collection_name + ".ref"].delete_one({"_id": id})
+
+
+def migrateData(data):
+  migratedData = []
+
+  for key in data:
+    migratedData.append({ "key": key, "value": data[key]})
+
+  return migratedData
+
+def updateActivities(coll, activities, sequenceId, parent = None):
+  for activity in activities:
+    updateProps =  {"$set": {"sequenceId": sequenceId, "startDate": activity["startDate"], "endDate": activity["endDate"] }, "$unset": {"parents": ""}}
+
+    if parent:
+      updateProps["$set"]["parent"] = uuid.UUID(parent)
+
+    query = {"_id": uuid.UUID(activity["id"])}
+
+    currentActivity = coll.find_one(query, {"resources": 1, "data":1})
+    if "resources" in currentActivity and type(currentActivity["resources"]) is list:
+      updateProps["$set"]["resources"] = { "shared_id": currentActivity["resources"] }
+
+    if "data" in currentActivity and type(currentActivity["data"]) is dict:
+      updateProps["$set"]["data"] = migrateData(currentActivity["data"])
+
+    coll.update_one(query, updateProps)
+
+    if "subTasks" in activity:
+      updateActivities(coll, activity["subTasks"], sequenceId, activity["id"])
+
+
+if len(sys.argv) < 6:
+    print("Not enough arguments.")
+    print("python " + os.path.basename(__file__) + " <mongoURL> <mongoPort> <userName> <password> <fsDirectory> [<dryRun>]")
+    print("<dryRun> is true by default")
+    sys.exit(0)
+
+mongoURL = sys.argv[1]
+mongoPort = sys.argv[2]
+userName = sys.argv[3]
+password = sys.argv[4]
+fsDirectory = sys.argv[5]
+dryRun = True if len(sys.argv) < 7 else sys.argv[6] != "false"
+
+if not dryRun:
+  print("dryRun = False: Commiting to database")
+else:
+  print("dryRun = True: Trial run")
+
+connString = "mongodb://"+ userName + ":" + password +"@"+mongoURL + ":" + mongoPort + "/"
+
+
+def updateActivitiesSchema(db, modelId, sequenceId):
+  activitiesCollName = modelId + ".activities"
+  fileContents = getFile(db, fsDirectory, activitiesCollName, str(sequenceId))
```

Comment:
instead of reading the gridFS list, isn't it better to just loop through the activities database?

From what I can tell, we're reading the cache file, then base on that take the ID and get the file and alter the schema.

And we ultimately ditch the activitiyList cache, so we're reading the file only to get the IDs?